cmd/integrationArtifactUpload.go

```diff
@@ -0,0 +1,205 @@
+		return errors.Wrapf(httpErr, "HTTP %v request to %v failed with error: %v", httpMethod, uploadIflowStatusURL, responseBody)
+	}
+	return errors.Errorf("Failed to create Integration Flow artefact, Response Status code: %v", uploadIflowStatusResp.StatusCode)
+}
+
+//UpdateIntegrationArtifact - Update existing integration artifact
+func UpdateIntegrationArtifact(config *integrationArtifactUploadOptions, httpClient piperhttp.Sender, fileUtils piperutils.FileUtils) error {
+	httpMethod := "POST"
+	header := make(http.Header)
+	header.Add("content-type", "application/json")
+	updateIflowStatusURL := fmt.Sprintf("%s/api/v1/IntegrationDesigntimeArtifactSaveAsVersion?Id='%s'&SaveAsVersion='%s'", config.Host, config.IntegrationFlowID, config.IntegrationFlowVersion)
+	payload, jsonError := GetJSONPayloadAsByteArray(config, "update", fileUtils)
+	if jsonError != nil {
+		return errors.Wrapf(jsonError, "Failed to get json payload for file %v, failed with error", config.FilePath)
+	}
+	updateIflowStatusResp, httpErr := httpClient.SendRequest(httpMethod, updateIflowStatusURL, payload, header, nil)
+
+	if updateIflowStatusResp != nil && updateIflowStatusResp.Body != nil {
+		defer updateIflowStatusResp.Body.Close()
+	}
+
+	if updateIflowStatusResp == nil {
+		return errors.Errorf("did not retrieve a HTTP response: %v", httpErr)
+	}
+
+	if updateIflowStatusResp.StatusCode == http.StatusOK {
+		log.Entry().
+			WithField("IntegrationFlowID", config.IntegrationFlowID).
+			Info("Successfully updated integration flow artefact in CPI designtime")
+		return nil
+	}
+	if httpErr != nil {
+		responseBody, readErr := ioutil.ReadAll(updateIflowStatusResp.Body)
+		if readErr != nil {
+			return errors.Wrapf(readErr, "HTTP response body could not be read, Response status code: %v", updateIflowStatusResp.StatusCode)
+		}
+		log.Entry().Errorf("a HTTP error occurred! Response body: %v, Response status code: %v", responseBody, updateIflowStatusResp.StatusCode)
+		return errors.Wrapf(httpErr, "HTTP %v request to %v failed with error: %v", httpMethod, updateIflowStatusURL, responseBody)
+	}
+	return errors.Errorf("Failed to update Integration Flow artefact, Response Status code: %v", updateIflowStatusResp.StatusCode)
+}
+
+//GetJSONPayloadAsByteArray -return http payload as byte array
+func GetJSONPayloadAsByteArray(config *integrationArtifactUploadOptions, mode string, fileUtils piperutils.FileUtils) (*bytes.Buffer, error) {
+	fileContent, readError := fileUtils.FileRead(config.FilePath)
+	if readError != nil {
+		return nil, errors.Wrapf(readError, "Error reading file")
+	}
+	jsonObj := gabs.New()
+	if mode == "create" {
+		jsonObj.Set(config.IntegrationFlowID, "Name")
+		jsonObj.Set(config.IntegrationFlowID, "Id")
+		jsonObj.Set(config.PackageID, "PackageId")
+		jsonObj.Set(b64.StdEncoding.EncodeToString(fileContent), "ArtifactContent")
+	} else {
```

Comment:
Now the "update" mode is silently ommitted. That means everything which is not "create" is automatically considered to be an "update". What I meant earlier was something like

```
if  mode == "create" {
  // blah
} else if mode == "update" {
  // blub
} else {
  return fmt.Errorf("Unkown node: '%s'", mode)
}
```

In case someone adjusts the code in the future to some new use-cases there is no surprise when we silently assume a new mode (e.g. "delete") as an "update".

Reply:
since API supported either create or update in this context which is decided from main function, so 3rd case not required i feel.

Reply:
I still have severe doubts that assuming "create" is create and any other string is "update" is a suitable approach.

> since API supported either [...]

That might be changed in the future and the change management process is more save in case we don't have implicit assumptions like described above.

> which is decided from main function [...]

Is it a valid approach to make an assumption how / from where a method is called? It can be called from everywhere or from everywhere in the same package in case it is private. Maybe the code needs to be refactored later e.g. in order to allow new use cases. In that case the current assumption how a function is called does not fit anymore.

In case you don't really share the concerns, keep it like it is (_it is basically your team which have to maintain it later_ :-)).

Reply:
fixed it in latest commit